core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -164,6 +172,26 @@ public void start(final Listener<RespT> observer, Metadata headers) {
     checkNotNull(observer, "observer");
     checkNotNull(headers, "headers");
 
+    // Create the context
+
+    // Same time used to compute all timeouts, in order to avoid inaccuracies and negative
+    // timeouts.
+    final long now = System.nanoTime();
+    final Deadline callDeadline = callOptions.getDeadlineNanoTime() != null
+        ? Deadline.after(callOptions.getDeadlineNanoTime() - now, NANOSECONDS)
+        : null;
+    final Deadline effectiveDeadline = min(callDeadline, parentContext.getDeadline());
+    if (effectiveDeadline != parentContext.getDeadline()) {
+      context = parentContext.withDeadline(effectiveDeadline, deadlineCancellationExecutor);
+    } else {
+      context = parentContext.withCancellation();
+    }
+
+    // TODO: Not nice. Improve. Maybe context should cancel itself immediately if deadline exceeded
+    if (context.getDeadline() != null && context.getDeadline().isExpired()) {
+      ((Context.CancellableContext) context).cancel(new TimeoutException("context timed out"));
```

Comment:
@louiscryan I might be missing something, but doesn't seem like Context gets canceled immediately if timeline is exceeded? See https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/Context.java#L659